static/js/services/firefoxextension.js

```diff
@@ -0,0 +1,113 @@
+			this.available = false;
+			this.e = $({});
+			this.autoinstall = {};
+			this.initialize();
+		};
+
+		FirefoxExtension.prototype.initialize = function() {
+			var marker = $window.document.getElementById("firefoxextension-available");
+			if (marker) {
+				this.available = true;
+				console.log("Firefox extension is available.");
+				this.e.triggerHandler("available", true);
+			} else if (!marker && this.available) {
+				this.available = false;
+				console.log("Firefox extension is no longer available.");
+				this.e.triggerHandler("available", false);
+			}
+		};
+
+		FirefoxExtension.prototype.registerAutoInstall = function(installFunc, cancelInstallFunc, force) {
+			this.autoinstall.install = installFunc;
+			this.autoinstall.cancel = cancelInstallFunc;
+			this.autoinstall.force = !!force;
+			if (!this.available && installFunc) {
+				this.e.triggerHandler("available", true);
+			}
+		};
+
+		var EXTENSION_DOM_ID = 'firefoxextension-available';
+		var intervalSecs = 50;
+		var intervalCount = 1;
+		var hasBeenInstalled = function() {
+			return $window.document.getElementById(EXTENSION_DOM_ID);
+		};
+
+		/**
+		 * Checks for availability of the Firefox extension by looking for the id which the extension
+		 * will append to the body of the document. Unfortunately there is no callback
+		 * API implemented by Firefox which will allow other domains to see if an
+		 * extension is installed using `InstallTrigger.install`. Only priviledged
+		 * domains may use the callback.
+		 *
+		 * @param {int} How long of a timespan the function should check for the extension install at intervalSecs interval rate
+		 * @return {promise}
+		 */
+		FirefoxExtension.prototype.detectInstalled = function(maxTimeout) {
+			var defer = $q.defer();
+			var that = this;
+
+			var intervalPromise = $interval(function() {
+				if (hasBeenInstalled()) {
+					console.log("Auto install success Firefox extension");
+					$interval.cancel(intervalPromise);
+					that.initialize();
+					defer.resolve("Auto install success Firefox extension");
+				} else if (intervalCount * intervalSecs >= maxTimeout) {
+					$interval.cancel(intervalPromise);
+					defer.reject("Timeout while waiting for extension to become available");
+				}
+				intervalCount++;
+			}, intervalSecs);
+
+			return defer.promise;
+		};
+
+		// Create extension api and wait for messages.
+		var extension = new FirefoxExtension();
+
+		// Always register default auto install which tells user that extension is required
+		// if screen sharing can only work with extension.
+		if ($window.webrtcDetectedBrowser === "firefox" && $window.webrtcDetectedVersion >= 36) {
+			extension.registerAutoInstall(function() {
+				var d = $q.defer();
+				alertify.dialog.alert(translation._("Screen sharing requires a browser extension. Please add the Spreed WebRTC screen sharing extension to Firefox and try again."));
+				d.reject("Manual extension installation required");
```

Comment:
Extension should be optional. It needs to work on spreed.me without the extension and without ever prompting for it.